### PR TITLE
Subtitles scraped, coverted to vtt and bundled

### LINKED ISSUE #877: Mediawiki video subtitles should be scraped too
See for example https://commons.wikimedia.org/wiki/File:Carcinoid_syndrome.webm. It has arabic subtitle. Article of WPEN integrates it https://en.wikipedia.org/wiki/Carcinoid_syndrome. Currently the subtitle links are in the HTML code, but the subtitles files are not downloaded and URL not rewritten.

### PR DESCRIPTION
**Pull request by bakshiutkarsha**
**DONE**: Subtitles are successfully getting scraped, converted(srt to vtt) and saved.

The problem here is that the subtitle file won't load locally(chrome doesn't allow), it only runs on server. I checked it by uploading my .vtt file on this 

> https://firebasestorage.googleapis.com/v0/b/words-cd928.appspot.com/o/test.html?alt=media&token=1dd510f2-d8ec-4e62-82b2-8ad04fb4a484

to be sure that it runs on server.

Here is the further link which states the same thing: 

> https://stackoverflow.com/questions/16092609/subtitles-not-showing-using-webvtt

The only other solution I found was to create the text track dynamically:

> https://gist.github.com/denilsonsa/aeb06c662cf98e29c379

@midik  @kelson42  Please suggest what should be done? Is it ok if the file does not work in local and on server?
should resolve #877

### CODE REVIEW

**Comment by kelson42** [comment]
@bakshiutkarsha What do you mean exactly with "won't load locally"? What is exactly the reproduction steps? What about Firefox?

**Comment by bakshiutkarsha** [comment]
Reproduction Steps:

1. Checkout issue/877 branch 
2. Create a articlelist.txt file which will have "User:Kelson/test"
3. Create a zim by running:
    npm start -- --mwUrl=https://en.wikipedia.org/ --adminEmail=kelson@kiwix.org  --articleList=./articlelist.txt 
4. Open the zim file in browser, you will see the "track" tag src has the correct file (you can check it by opening it in a new tab).
5. You will observe the file is correct but still the subtitles are not showing.

Also, the same .vtt file works when I tried hosting it here:
https://firebasestorage.googleapis.com/v0/b/words-cd928.appspot.com/o/test.html?alt=media&token=1dd510f2-d8ec-4e62-82b2-8ad04fb4a484

No, it doesn't work locally on any browser (tried both on chrome and firefox).
For reference,  https://stackoverflow.com/questions/16092609/subtitles-not-showing-using-webvtt

**Comment by kelson42** [comment]
>     4\. Open the zim file in browser

How to you do that? With Kiwix JS or using Kiwix Serve?

**Comment by bakshiutkarsha** [comment]
> > ```
> > 4\. Open the zim file in browser
> > ```
> 
> How to you do that? With Kiwix JS or using Kiwix Serve?

I tried with both using kiwix-serve and kiwix plugin(chrome extension)

**Comment by kelson42** [comment]
> > > ```
> > > 4\. Open the zim file in browser
> > > ```
> > 
> > 
> > How to you do that? With Kiwix JS or using Kiwix Serve?
> 
> 
> I tried with both using kiwix-serve and kiwix plugin(chrome extension)

OK, please don't use KiwixJS to tests ZIM files. The way how browsers allow (or not) things in their extension is complex and cumbersom. The way how we deal with them is in Kiwix JS is exactly the same. Use only Kiwix-serve.

**Comment by kelson42** [comment]
@bakshiutkarsha I had a look. Please write the automated testing for this piece of code. So far I can see, the URL rewritting is wrong. BTW, never ever put an absolute URL in a ZIM file HTML content.

**Comment by bakshiutkarsha** [comment]
@midik The Travis CI is failing due to following reason:
The changes I have done in this PR is to download the subtitles, so it is trying to download subtitle from the article "Michael_Jackson" with this api:

https:///api.php?action=timedtext&title=File%3AMichael_Jackson_Thriller.ogg&lang=en&trackformat=srt  and getting the 

> Error: connect ECONNREFUSED 127.0.0.1:443

And because this article is in testcase from ages, no one has replaced it. So, do we change the article list?? @kelson42

**Comment by bakshiutkarsha** [comment]
@midik The travis build is failing because it is trying to download the subtitles from an article "Michael_Jackson", which is in our test case and because the "src" I am getting is:

> /w/api.php?action=timedtext&amp;title=File%3AMichael_Jackson_Thriller.ogg&amp;lang=en&amp;trackformat=srt

has no domain, it is trying to hit the API with this given API, thus the error. Also, even with the domain name, the **file doesn't seems to exists**. I assume this is an old article and no one has updated it recently.

@kelson42 what should we be doing in such cases? Do we rely that the src will have the correct API url?

**Comment by kelson42** [comment]
@bakshiutkarsha You should fix the `src` attribute value with the info you have to get a proper URL to request. The `src` value is proper. See:
```
$ curl -I "http://en.wikipedia.org/w/api.php?action=timedtext&title=File%3AMichael_Jackson_Thriller.ogg&lang=en&trackformat=srt"
HTTP/1.1 301 TLS Redirect
Date: Mon, 04 May 2020 17:06:08 GMT
Server: Varnish
X-Varnish: 557879365
X-Cache: cp3050 int
X-Cache-Status: int-front
Server-Timing: cache;desc="int-front"
Set-Cookie: WMF-Last-Access=04-May-2020;Path=/;HttpOnly;secure;Expires=Fri, 05 Jun 2020 12:00:00 GMT
Set-Cookie: WMF-Last-Access-Global=04-May-2020;Path=/;Domain=.wikipedia.org;HttpOnly;secure;Expires=Fri, 05 Jun 2020 12:00:00 GMT
X-Client-IP: 2a02:168:61bf:0:7c59:480d:c195:1e4f
Location: https://en.wikipedia.org/w/api.php?action=timedtext&title=File%3AMichael_Jackson_Thriller.ogg&lang=en&trackformat=srt
Content-Length: 0
Connection: keep-alive

```

File: package.json [hunk 1]
```diff
@@ -97,6 +97,7 @@
     "semver": "^5.6.0",
     "service-mobileapp-node": "git+https://github.com/wikimedia/mediawiki-services-mobileapps.git#d244439",
     "service-runner": "^2.7.2",
+    "srt-to-vtt": "^1.1.3",
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply package.json:100]: `.srt` is does not work with html5.
Also, according MDN, the src should be a `.vtt` file, there is no mention of `.srt`.
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/track

For more info,
https://stackoverflow.com/questions/26968759/does-html5-video-support-srt-texttrack

File: test/unit/saveArticles.test.ts [hunk 2]
```diff
@@ -237,3 +239,33 @@ test('--customFlavour', async (t) => {
     t.ok(ParisDocument.querySelector('#PRE_PROCESSOR'), `Paris was correctly pre-processed`);
     t.ok(PragueDocument.querySelector('#POST_PROCESSOR'), `Prague was correctly post-processed`);
 });
+
+
+test('treat article with Subtitles', async (t) => {
+    const { downloader, mw, dump, zimCreator } = await setupScrapeClasses({}, true)
+
+    const subtitleExists = await checkSubtitles('User:Kelson/test', downloader, dump, zimCreator, mw);
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply test/unit/saveArticles.test.ts:247]: Resolved

File: src/util/saveArticles.ts [hunk 3]
```diff
@@ -405,9 +409,42 @@ async function treatVideo(mw: MediaWiki, dump: Dump, srcCache: KVS<boolean>, art
...
 }
 
+export async function treatSubtitles(videoEl: any, articleId: string, downloader: Downloader, zimCreator: ZimCreator, webUrlHost: string, mw: MediaWiki) {
+    const trackEle = videoEl.querySelector('track');
+    if (trackEle) {
+        const sourceUrl = getFullUrl(webUrlHost, trackEle.getAttribute('src'), mw.base);
+        const trackTitle: any =  QueryStringParser.parse(sourceUrl).title;
+        const { content }  = await downloader.downloadContent(sourceUrl);
+        return new Promise((resolve, reject) => {
+            fs.writeFile(`/tmp/${trackTitle}.srt`, content.toString(), 'utf8', function (err: any) {
```
[the remark this answers is not in the document]
Comment (bakshiutkarsha) [reply src/util/saveArticles.ts:426]: srt-to-vtt package that I have used, needs filesystem to be involved.

https://www.npmjs.com/package/srt-to-vtt

**Comment by kelson42** [comment]
I don't feel comfortable at all having us converting SRT to VTT. This something which can go wrong and not our role. On the top of that:
* The mediawiki team is on its way to support/move to VTT (for the same reason like us) https://phabricator.wikimedia.org/T116154
* We will introduce oggjs/videojs (which are able to read srt) #597

I propose to keep the code to rewrite the URL and just skip any node which is not in VTT format (which for the moment will probably be all of them).

**Comment by kelson42** [comment]
@bakshiutkarsha Do you have seen that Mediawiki seems to provide these same subtitle in VTT as well? see for example: "https://commons.wikimedia.org/w/api.php?action=timedtext&title=File%3ACarcinoid_syndrome.webm&lang=ar&trackformat=vtt"

**Comment by bakshiutkarsha** [comment]
> @bakshiutkarsha Do you have seen that Mediawiki seems to provide these same subtitle in VTT as well? see for example: "https://commons.wikimedia.org/w/api.php?action=timedtext&title=File%3ACarcinoid_syndrome.webm&lang=ar&trackformat=vtt"

No, I wasn't aware of that. 
I have done the changes, where the `trackformat` will be replaced to `vtt` always. Hope this is the right solution.

**Pull request closed without merging**